File: Algorithm_111_Minimum Depth of Binary Tree.cpp
#include <iostream>
#include <algorithm>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
int minDepth(TreeNode* root)
{
	if (root==NULL) return 0;
	else if (root->left==NULL && root->right!=NULL) return minDepth(root->right)+1;
	else if (root->left!=NULL && root->right==NULL) return minDepth(root->left)+1;
	else return min(minDepth(root->left),minDepth(root->right))+1;
}
int main()
{
	TreeNode *t1=new TreeNode(1);
	TreeNode *t2=new TreeNode(2);
	t1->left=t2;
	cout<<minDepth(t1);
	return 0;
}

File: Algorithm_111_Minimum Depth of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// 思路：直接递归，用变量minDepth来记录最小深度值，递归中每当遇到叶子节点时，就尝试更新minDepth

class Solution {

    private int minDepth = Integer.MAX_VALUE;

    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        getDepth(root, 1);
        return minDepth;
    }

    private void getDepth(TreeNode root, int curDepth) {
        if (root == null) {
            return;
        }
        if (root.left == null && root.right == null) { // root is leaf
            minDepth = Math.min(minDepth, curDepth);
        }
        getDepth(root.left, curDepth + 1);
        getDepth(root.right, curDepth + 1);
    }

}